gpii/node_modules/flowManager/src/SolutionsRegistryDataSource.js

```diff
@@ -35,45 +43,197 @@ fluid.defaults("gpii.flowManager.solutionsRegistry.dataSource", {
+    members: {
+        repositorySolutionsRegistry: null
+    },
+    invokers: {
+        getImpl: {
+            funcName: "gpii.flowManager.solutionsRegistry.dataSource.local.handle",
+            args: ["{that}", "{arguments}.1"]
+                             // options
+        }
+    },
+    listeners: {
+        "loadSolutions.loadFromLocalDisk": {
+            listener: "{that}.loadFromLocalDisk",
+            priority: "first"
+        },
+        "loadSolutions.getRevision": {
+            listener: "{revisionRequester}.getRevision",
+            priority: "after:loadFromLocalDisk"
+        },
+        "loadSolutions.loadFromRepository": {
+            listener: "gpii.flowManager.solutionsRegistry.dataSource.local.loadFromRepository",
+            args: ["{that}", "{arguments}.0", "{that}.platformReporter"],
+                             // revision
+            priority: "after:getRevision"
+        },
+        "loadSolutions.solutionsLoaded": {
+            listener: "{that}.events.solutionsRegistryReady",
+            priority: "after:loadFromRepository"
+        }
+    }
+});
+
+/**
+ * Retrieve the solutions registry JSON file from the source code repository.
+ * If successful, sets the repositorySolutionsRegistry member to the fetched
+ * solutions registry.
+ * @param {Component} that - An instance of gpii.flowManager.solutionsRegistry.dataSource
+ * @param {Component} that.repositorySolutionsLoader - An instance of
+ *                    gpii.flowManager.repositorySolutionsLoader.
+ * @param {Object} that.repositorySolutionsRegistry - Set to point to the
+ *                 solutions registry retrieved from the respository, or null if
+ *                 none were retrieved.
+ * @param {Object} revision - the SHA256 revision of the repository in the form:
+ *                            { sha256: ... }.  If the sha256 is missing, the
+ *                            result of the load is null.
+ * @param {Component} platformReporter - used to get the name of the platform,
+ *                                       e.g., "win32".
+ * @return {Promise} A promise whose resolved value is eiher the solutions
+ *      registry, or, if there is an error, an object with an "isError: true"
+ *      property.
+ */
+gpii.flowManager.solutionsRegistry.dataSource.local.loadFromRepository = function (that, revision, platformReporter) {
+    var gpiiRevision = revision.sha256;
+    var platformId = platformReporter.reportPlatform().id;
+    var fileName = platformId + ".json5";
+    var togo = that.repositorySolutionsLoader.getSolutions(gpiiRevision, fileName, platformId);
+    togo.then(function () {
+        // Either the solutions registry has been successfully retrieved from the
+        // repository, or it hasn't (= null).  Set that.repositorySolutionsRegistry
+        // to whatever the result.
+        that.repositorySolutionsRegistry = that.repositorySolutionsLoader.solutionsRegistry;
```

Comment:
It probably makes more sense to have `that.repositorySolutionsLoader.getSolutions` return a promise whose resolved value is the expected solutions registry, rather than reaching into `that.repositorySolutionsLoader.solutionsRegistry`.

Reply:
Okay.  But as I fixed it, I noticed that the RepositorySolutionLoader was using `resolve()` for the error conditions.  I've switched that to `reject()`, and adjusted its tests and the SolutionsRegistryDataSource appropriately.